Significant-Gravitas/Auto-GPT
# Make prompt parameters configurable

Loading prompt constraints, resources and performance evaluation from a yaml file (default prompt_settings.yaml) 
Can be set from .env (PROMPT_SETTINGS_FILE) or by commandline (--prompt-settings or -P)

potentially resolves #3954

### Background
The main reason for proposed this changes is because they can help with with different LLM models, we talked about that in #25 #567 #2158.
They don't handle the prompts in the same way as GPT3.5/GPT4, and they often get confused. this way can be easy to create and share prompts made specifically for them.

Also, it can be useful for models made for different languages, like for example https://github.com/FreedomIntelligence/LLMZoo

....Or just for people who want to have more control on what AutoGPT can do without having to modify the code

### Changes
Moved the hardcoded default prompt constraints, resources and performance evaluation form prompt.py to the new file prompt_settings.yaml, wich will be used as default file.

Added the new configuration variable PROMPT_SETTINGS_FILE in .env.template and modified config.py to handle it

Added the new file autogpt/config/prompt_config.py, wich contain the PromptConfig class, wich is initialized passing the file path and contains the datas from the configuration file

Moved the hardcoded default prompt constraints, resources and performance evaluation form prompt.py to the new file prompt_settings.yaml, wich will be used as default file.

Modified prompt.py to use the values from the PromptConfig instead of hardcoded datas

Modified cli.py, main.py and configurator.py to handle the new --prompt-settings / -P commandline args

### Documentation
The new .env variable PROMPT_SETTINGS_FILE is described there, while the new --prompt-settings/-P comd line args are described both in cli.py and in usage.md.
I followed the same policy used for the ai.settings.yaml file

### Test Plan
- Start AutoGPT without modifying any configurations, should work just as before
- Start AutoGPT using --prompt-settings (file) (ex. python -m autogpt -P prompt_settings_ex.yaml), where file doesn't exists or isn't valid. AutoGPT should give a validation error and stop
- Start AutoGPT using after setting PROMPT_SETTINGS_FILE=(file), where file doesn't exists or isn't valid. AutoGPT should give a validation error and stop
- Copy the prompt_settings.yaml file and change it a bit, while keeping it still valid. Run AutoGPT normally, it should still run as expected

To check if the prompt have actually changed, i also used those changes in my fork (see #2594) while connecting to https://github.com/keldenl/gpt-llama.cpp
I know it isn't something officially supported, but it's still a good and quick way to see what's going on, since the webservice prints the prompt on the standard output

### PR Quality Checklist
- * [x] My pull request is atomic and focuses on a single change.
- * [x] I have thoroughly tested my changes with multiple different prompts.
- * [x] I have considered potential risks and mitigations for my changes.
- * [x] I have documented my changes clearly and comprehensively.
- * [x] I have not snuck in any "extra" small tweaks changes
